### ISSUE TITLE
LinqToDBConvertException when materializing entity

### ISSUE THREAD

**Issue by genusP** [linq2db/linq2db#975]
linq2db.core: 1.10.1
Query
```C#
var employeeId =10;
(from t in db.GetTable<Task>()
                     join a in db.GetTable<Assignment>()
                         on new { t.DirectionId, t.TargetId, t.TargetName }
                         equals new { a.DirectionId, a.TargetId, a.TargetName }
                     where a.EmployeeId == employeeId
                         && (a.DateRevoke == null || a.DateRevoke > SqlServer.GetDate())
                     select t)
.Distinct()
.Where(it => it.ActualStage.Any(d => d.StageId < 9000 || ((int?)d.StageId) == null));
```
Exception
```
LinqToDB.Common.LinqToDBConvertException: Cannot convert value 'c5c0a778-694e-49d1-b1a0-f8ef5569c673' to type 'System.Int32' ---> System.InvalidCastException: Object must implement IConvertible.
at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
at LinqToDB.Common.ConvertBuilder.ConvertDefault(Object value, Type conversionType)
--- End of inner exception stack trace ---
at LinqToDB.Common.ConvertBuilder.ConvertDefault(Object value, Type conversionType)
at lambda_method(Closure , IQueryRunner , IDataReader )
at LinqToDB.Linq.QueryRunner.Mapper`1.Map(IQueryRunner queryRunner, IDataReader dataReader)
at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__8`1.MoveNext()
```

Models
```c#
[Table(Schema="awt", Name="Tasks")]
	public partial class Task : ITargetAware
	{
		[PrimaryKey, Identity   ] public int       Id          { get; set; } // Int
		[Column,     NotNull    ] public DateTime  DateBegin   { get; set; } // DateTime
		[Column,        Nullable] public DateTime? DateEnd     { get; set; } // DateTime
		[Column,     NotNull    ] public Guid      DirectionId { get; set; } // UniqueIdentifier
		[Column,     NotNull    ] public string    Text        { get; set; } // varchar(1000)
		[Column,     NotNull    ] public string    TargetName  { get; set; } // nvarchar(128)
		[Column,     NotNull    ] public int       TargetId    { get; set; } // Int
		[Column,        Nullable] public int?      ParentId    { get; set; } // Int

                [Association(ThisKey =nameof(Id), OtherKey =nameof(TaskStage.TaskId), ExpressionPredicate = nameof(ActualTaskExp), CanBeNull =true)]
                public IEnumerable<TaskStage> ActualStage { get; set; }

                private static Expression<Func<Task, TaskStage, bool>> ActualTaskExp()
                        => (t, ts) => ts.Actual == true;
      }

[Table(Schema="awt", Name="Assignments")]
	public partial class Assignment : ITargetAware
	{
		[PrimaryKey, Identity   ] public int       Id          { get; set; } // Int
		[Column,     NotNull    ] public int       EmployeeId  { get; set; } // Int
		[Column,     NotNull    ] public DateTime  DateAssign  { get; set; } // DateTime
		[Column,        Nullable] public DateTime? DateRevoke  { get; set; } // DateTime
		[Column,     NotNull    ] public Guid      DirectionId { get; set; } // UniqueIdentifier
		[Column,     NotNull    ] public string    TargetName  { get; set; } // nvarchar(128)
		[Column,     NotNull    ] public int       TargetId    { get; set; } // Int
       }

public interface ITargetAware
    {
        string TargetName { get; }
        int TargetId { get; }
    }

[Table(Schema="awt", Name="TaskStages")]
	public partial class TaskStage
	{
		[PrimaryKey, Identity   ] public int      Id           { get; set; } // Int
		[Column,     NotNull    ] public int      TaskId       { get; set; } // Int
		[Column,     NotNull    ] public int      StageId      { get; set; } // Int
		[Column,     NotNull    ] public DateTime SwitchDate   { get; set; } // DateTime
		[Column,     NotNull    ] public bool     Actual       { get; set; } // Bit
		[Column,        Nullable] public int?     WorkReportId { get; set; } // Int
		[Column,        Nullable] public int?     UserId       { get; set; } // Int
}

public static class SqlServer
    {
        [Sql.Function(ServerSideOnly = true)]
        public static DateTime GetDate() { throw new InvalidOperationException("Use only LINQ expression"); }
    }
```
SQL
```sql
DECLARE @employeeId Int -- Int32
SET     @employeeId = 10

SELECT
	*
FROM
	(
		SELECT DISTINCT
			[t].[Id],
			[t].[DateBegin],
			[t].[DateEnd],
			[t].[DirectionId],
			[t].[Text],
			[t].[TargetName],
			[t].[TargetId],
			[t].[ParentId]
		FROM
			[awt].[Tasks] [t]
				INNER JOIN [awt].[Assignments] [a] ON [t].[DirectionId] = [a].[DirectionId] AND [t].[TargetId] = [a].[TargetId] AND [t].[TargetName] = [a].[TargetName]
		WHERE
			[a].[EmployeeId] = @employeeId AND ([a].[DateRevoke] IS NULL OR [a].[DateRevoke] > GetDate())
	) [it]
WHERE
	EXISTS(
		SELECT
			*
		FROM
			[awt].[TaskStages] [t1]
		WHERE
			[it].[Id] = [t1].[TaskId] AND [t1].[Actual] = 1 AND
			([t1].[StageId] < 9000 OR [t1].[StageId] IS NULL)
	)
```

**Comment by sdanyliv** [comment]
@genusP, missed `ITargetAware` and `TaskStage`. Can you update issue description with these entities?

**Comment by genusP** [comment]
```c#
public interface ITargetAware
    {
        string TargetName { get; }
        int TargetId { get; }
    }

[Table(Schema="awt", Name="TaskStages")]
	public partial class TaskStage
	{
		[PrimaryKey, Identity   ] public int      Id           { get; set; } // Int
		[Column,     NotNull    ] public int      TaskId       { get; set; } // Int
		[Column,     NotNull    ] public int      StageId      { get; set; } // Int
		[Column,     NotNull    ] public DateTime SwitchDate   { get; set; } // DateTime
		[Column,     NotNull    ] public bool     Actual       { get; set; } // Bit
		[Column,        Nullable] public int?     WorkReportId { get; set; } // Int
		[Column,        Nullable] public int?     UserId       { get; set; } // Int
}
```

**Comment by sdanyliv** [comment]
Reproduced, thanks!

**Comment by genusP** [comment]
if remove last Where, code work correctly.

**Comment by sdanyliv** [comment]
@genusP, we are working on fix.
Simple workaround that i’ve found to replace `join` with our extension [InnerJoin](https://github.com/linq2db/linq2db/wiki/Join-Operators#using-innerjoin-function)

**Comment by genusP** [comment]
Workaround helped. Thanks.

**Comment by sdanyliv** [comment]
Fixed in #1197
